### PR TITLE
Prevent instructions to be generated as names.

### LINKED ISSUE #5229: [Yul] Avoid clashes with builtin functions
The disambiguator and the name dispenser should never generate names that match the name of builtin functions. This includes names of instructions and builtin functions of the dialect.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/5229

### CODE REVIEW

File: libyul/AsmParser.h [hunk 1]
```diff
@@ -46,6 +46,8 @@ class Parser: public langutil::ParserBase
 	/// @returns an empty shared pointer on error.
 	std::shared_ptr<Block> parse(std::shared_ptr<langutil::Scanner> const& _scanner, bool _reuseScanner);
 
+	static std::map<std::string, dev::eth::Instruction> const& instructions();
```
Comment (erak) [inline libyul/AsmParser.h:50]: I think a comment would be good.

File: libyul/optimiser/VarNameCleaner.cpp [hunk 2]
```diff
@@ -108,7 +110,11 @@ YulString VarNameCleaner::findCleanName(YulString const& _name) const
 
 bool VarNameCleaner::isUsedName(YulString const& _name) const
 {
-	return m_dialect.builtin(_name) || m_usedNames.count(_name);
+	if (m_dialect.builtin(_name) || m_usedNames.count(_name))
```
Comment (erak) [inline libyul/optimiser/VarNameCleaner.cpp]: Is `_name` allowed to be empty or do we also need `_name.empty()`? If so, can this be shared with `NameDispenser::illegalName()`?

Comment (chriseth) [reply libyul/optimiser/VarNameCleaner.cpp]: I think it's difficult to share the code there, but I can move the empty check here.

Comment (chriseth) [reply libyul/optimiser/VarNameCleaner.cpp]: Also note that the EVMDialect stuff below will be removed once we can move the EVM instructions to proper builtins.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**